Implementing memory

tischi opened the pull request:
Builds on https://github.com/haesleinhuepf/bia-tischi/pull/10

tischi committed:
WIP implementing memory

tischi committed:
Memory working, need to implement token count next

tischi committed:
Improve chat history printing

tischi committed:
Add total session cost information

tischi commented:
I think this could be merged now.

haesleinhuepf commented on src/bia_tischi/_utilities.py:
```diff
@@ -57,16 +57,79 @@ def find_variable(variables, key, type_checker_function):
+    return code, text
+
+
+def create_pricing_summary(model, chat_history, full_response, system_prompt, user_prompt):
+    import tiktoken
+    from ._machinery import Models, Context
+    encoding = tiktoken.encoding_for_model(model)
+    input_token = (len(encoding.encode(user_prompt))
+                   + len(encoding.encode(system_prompt))
+                   + len(encoding.encode(concatenate_chat_content(chat_history))))
+    output_token = len(encoding.encode(full_response))
+    pricing = "\n##### Pricing\n\n"
+    pricing += "- Model: " + model + "\n"
+    pricing += "- Pricing: https://openai.com/pricing\n"
+    pricing += "- Last input: " + str(input_token) + " token = "
+    input_price = Models.usd_per_1k_input_token(model) * input_token / 10.0
+    pricing += print_costs_in_cent(input_price)
+    pricing += "- Last output: " + str(output_token) + " token = "
+    output_price = Models.usd_per_1k_output_token(model) * output_token / 10.0
+    pricing += print_costs_in_cent(output_price)
+    Context.session_price_us_cent += input_price + output_price
+    pricing += "- Total session costs: " + print_costs_in_cent(Context.session_price_us_cent)
+    pricing += "\n"
+    return pricing
```
Out of curiosity: is this related to implementing _memory_? I would leave the pricing thing out for now as it might be a research topic itself.

tischi replied:
no it is not, I did bunch of stuff in parallel

haesleinhuepf commented on setup.cfg:
```diff
@@ -39,6 +39,7 @@ install_requires =
     matplotlib
     scikit-learn
     tiktoken
+    langchain
```
I do not see code which uses langchain (except in a notebook). Is langchain really a dependency?

tischi replied:
no, we can remove it

haesleinhuepf reviewed (commented):
I'm struggling a bit to review this because changes related to memory are mixed with changes in other contexts (renaming functions, adding pricing, ...). I'm ok merging this. I'm just a big fan of concise PRs with only related changes. ;-)

tischi commented:
> I'm struggling a bit to review this because changes related to memory are mixed with changes in other contexts (renaming functions, adding pricing, ...). I'm ok merging this. I'm just a big fan of concise PRs with only related changes. ;-)

Yes, you are right, but it was so much that I changed that I did not see how to do this with individual PRs. I refactored almost everything....
